code/datums/music/music.dm

```diff
@@ -0,0 +1,263 @@
+	if(is_fading)
+		if(fade_volume_target > fade_volume)
+			fade_volume = min(fade_volume + wait * fade_rate, fade_volume_target)
+		else if(fade_volume_target < fade_volume)
+			fade_volume = max(fade_volume - wait * fade_rate, fade_volume_target)
+		else
+			set_is_fading(FALSE)
+		if(sound_datum && target)
+			sound_datum.status = SOUND_UPDATE
+			sound_datum.volume = fade_volume * base_volume
+			SEND_SOUND(target, sound_datum)
+		if(fade_volume == 0 && fade_volume_target == 0)
+			qdel(src)
+
+/datum/music/proc/set_is_fading(new_is_fading)
+	new_is_fading = !!new_is_fading
+	if(new_is_fading == is_fading)
+		return
+	is_fading = new_is_fading
+	if(is_fading)
+		STOP_PROCESSING(SSprocessing, src)
+		START_PROCESSING(SSfastprocess, src)
+	else
+		STOP_PROCESSING(SSfastprocess, src)
+		START_PROCESSING(SSprocessing, src)
+
+
+/datum/music/sourced
+	var/mob/target_mob
+	var/list/datum/component/music_player/players = list()
+	/// Range within which music will play at full volume
+	var/full_range = 10
+	/// Maximum range within which music will play, out of that range it will stop
+	var/soft_range = 20
+	/// Speed at which music will fade out while out of range
+	var/range_fade_speed = 0.1
+
+/datum/music/sourced/New(client/C, _fade_volume)
+	if(C)
+		target_mob = C.mob
+	..()
+
+/datum/music/sourced/process()
+	if(!target_mob || !target_mob.client)
+		qdel(src)
+		return
+	var/closest_range_squared = soft_range*soft_range + 10
+	for(var/_player in players)
+		var/datum/component/music_player/player = _player
+		var/atom/movable/M = player.parent
+		if(M.z != target_mob.z)
+			continue
+		var/dx = M.x-target_mob.x
+		var/dy = M.y-target_mob.y
+		var/range_squared = dx*dx+dy*dy
+		if(range_squared < closest_range_squared)
+			closest_range_squared = range_squared
+	var/closest_range = sqrt(closest_range_squared)
+	var/target
+	if(closest_range <= full_range)
+		target = 1
+	else if(closest_range >= soft_range)
+		target = 0
+	else
+		target = (soft_range - closest_range) / (soft_range - full_range)
+	if(!is_fading || fade_volume_target != target)
+		fade_at_rate(target, range_fade_speed)
+
+	. = ..()
+
+/datum/music/sourced/Destroy()
+	. = ..()
+	for(var/_player in players)
+		var/datum/component/music_player/player = _player
+		player.mob_players -= target_mob
+
+/datum/component/music_player
+	var/list/mob_players = list()
+	var/music_path
+	/// Whether new mobs will be added. To force mobs to stop playing this, use stop_all()
+	var/enabled = TRUE
+	/// Range within which music will start to play
+	var/start_range = 7
+	var/fade_in_time = 20
+	/// Whether music of the same typepath is shared
+	var/shared = TRUE
+
+/datum/component/music_player/Initialize(_music_path)
+	if(_music_path)
+		music_path = _music_path
+	START_PROCESSING(SSprocessing, src)
+
+/datum/component/music_player/proc/do_range_check(var/fade_time = fade_in_time)
+	if(!music_path)
+		return
+	var/shared = FALSE
+	for(var/mob/living/M in range(start_range, parent))
```

Comment:
The music should play to ghosts nearby too

Reply:
I intentionally made it not play to ghosts since that would make it play when the fight isn't actually happening